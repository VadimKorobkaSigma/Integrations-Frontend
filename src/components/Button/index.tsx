import * as React from 'react';
import cn from 'classnames';

import styles from './styles.module.scss';

interface Props {
    children: string | JSX.Element | JSX.Element[];
    className?: string;
    type?: 'submit';
    onClick?: React.DOMAttributes<HTMLButtonElement>['onClick'];
}

const Button: React.FC<Props> = ({ children, ...args }) => {
    return (
        <button {...{ ...args }} className={cn(styles.button, args.className)}>
            {children}
        </button>
    );
};

export default React.memo(Button);
