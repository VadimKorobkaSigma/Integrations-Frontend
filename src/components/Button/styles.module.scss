.button {
  color: #fff;
  background-color: #2991f3;
  margin-bottom: 1rem;
  width: 160px;
  border: 0;
  padding: 0.5rem 1rem;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.15);
  border-radius: 4px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  font-size: 20px;

  &:hover {
    background-color: #1579d7;
  }
}
