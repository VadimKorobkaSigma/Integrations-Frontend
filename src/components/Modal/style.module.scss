.modal {
  position: absolute;
  z-index: 1;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;

  .backdrop,
  .container {
    transition: opacity 0.6s;
    opacity: 0;
  }

  &.open {
    pointer-events: initial;

    .backdrop,
    .container {
      opacity: 1;
    }
  }

  .backdrop {
    background-color: #00000060;
    position: absolute;

    width: 100%;
    height: 100%;
  }

  .container {
    position: absolute;
    top: 10%;
    bottom: 10%;
    left: 25%;
    right: 25%;
    color: $text-black;
    background-color: $white;
    box-shadow: #00000012 0px 4px 10px;
    padding: 24px 32px;
    border-radius: 4px;
    display: grid;
    grid-template-rows: fit-content(0) 1fr fit-content(0);
    border-top: 4px solid #2991f3;
  }
}

.header {
  h1 {
    text-align: center;
    color: $grey-dark;
  }
}

.header {
  text-align: center;
  color: $grey-dark;
}

.footer {
  display: flex;
  justify-content: flex-end;

  & > *:not(:last-child) {
    margin-right: 1rem;
  }
}
