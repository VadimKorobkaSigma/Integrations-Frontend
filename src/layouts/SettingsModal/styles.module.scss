.modal {
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  padding: 3rem;
  box-shadow: rgba(0, 0, 0, 0.12) 0px 4px 10px;
  border-radius: 4px;
  color: #565360;
  background: #fff;

  h1 {
    text-align: center;
    color: #616161;
  }

  > :not(:last-child) {
    margin-bottom: 1rem;
  }

  .submit {
    width: initial;
    margin-left: auto;
  }
}
