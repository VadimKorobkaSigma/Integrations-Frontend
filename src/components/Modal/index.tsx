import * as React from 'react';
import cn from 'classnames';

import styles from './style.module.scss';
import Button from '@components/Button';

interface Props {
    isOpen: boolean;
    header?: string | React.ReactElement;
    footer?: string | React.ReactElement;
    onClose: VoidFunction;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    children: React.ReactElement<any, any> | React.ReactNodeArray;
}

const Modal: React.FC<Props> = ({ children, isOpen, onClose, header, footer }) => {
    return (
        <div className={cn(styles.modal, isOpen && styles.open)}>
            <div className={styles.backdrop} onClick={onClose} />
            <div className={styles.container}>
                <header className={styles.header}>
                    <h1>{header}</h1>
                </header>
                {children}
                {footer && (
                    <footer className={styles.footer}>
                        <Button variant="link" onClick={onClose}>
                            Cancel
                        </Button>
                        {footer}
                    </footer>
                )}
            </div>
        </div>
    );
};

export default React.memo(Modal);
