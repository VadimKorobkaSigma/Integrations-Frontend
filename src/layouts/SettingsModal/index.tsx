import * as React from 'react';
import Modal from 'react-modal';
import styles from './styles.module.scss';
import Button from '@components/Button';
import api from '@services/api';
import useError from '@services/hooks/useError';
import Organization from '@dtos/organization';
import ErrorComponent from '@components/ErrorComponent';

Modal.setAppElement('#react');

interface OwnProps {
    selectedOrg: Organization | null;
    setSelectedOrg: React.Dispatch<React.SetStateAction<Organization | null>>;
}

type Props = OwnProps;

const SettingsModal: React.FC<Props> = ({ selectedOrg, setSelectedOrg }) => {
    const [error, handleError, clearError] = useError();
    const [secret, setSecret] = React.useState('');
    const [team, setTeam] = React.useState('');

    React.useEffect(() => {
        if (selectedOrg) {
            api.getSettings(selectedOrg.id)
                .then((result) => {
                    setSecret(result.cxgoSecret);
                    setTeam(result.team);
                    clearError();
                })
                .catch(handleError);
        } else {
            setSecret('');
            setTeam('');
        }
    }, [selectedOrg]);

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault();
        if (selectedOrg) {
            api.setSettings(selectedOrg.id, { team, cxgoSecret: secret })
                .then(() => {
                    setSelectedOrg(null);
                    clearError();
                })
                .catch(handleError);
        }
    };

    return (
        <Modal isOpen={!!selectedOrg} onRequestClose={() => setSelectedOrg(null)} className={styles.modal}>
            <form onSubmit={handleSubmit}>
                <h1>Settings of {selectedOrg?.name} organization</h1>
                <ErrorComponent error={error} />
                <label htmlFor="secret">Checkmarx Go Secret</label>
                <textarea placeholder="Secret" id="secret" value={secret} onChange={(e) => setSecret(e.target.value)} />
                <label htmlFor="team">Team</label>
                <textarea placeholder="Team" id="team" value={team} onChange={(e) => setTeam(e.target.value)} />
                <Button className={styles.submit} type="submit">
                    Submit
                </Button>
            </form>
        </Modal>
    );
};

export default React.memo(SettingsModal);
