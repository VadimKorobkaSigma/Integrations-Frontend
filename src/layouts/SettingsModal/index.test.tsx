import * as React from 'react';
import SettingsModal from './index';
import { mount } from 'enzyme';
import Modal from '@components/Modal';
import Button from '@components/Button';

const data = [
    { id: 'test_id', name: 'test 1', webhookEnabled: false },
    { id: 'test_id_2', name: 'test 2', webhookEnabled: true, webhookId: 1 },
];

test('SettingsModal Component', async () => {
    const setSelectedOrg = jest.fn();
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            ok: true,
            json: () => Promise.resolve(data),
        }),
    );
    jest.spyOn(React, 'useState').mockImplementation(
        jest
            .fn()
            .mockReturnValueOnce(['', jest.fn()])
            .mockReturnValueOnce(['Test secret', jest.fn()])
            .mockReturnValueOnce(['Test team', jest.fn()])
            .mockReturnValue([null, jest.fn()]),
    );
    const wrapper = mount(
        <SettingsModal selectedOrg={{ id: 'test_id', name: 'test 1' }} setSelectedOrg={setSelectedOrg} />,
    );

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find(Modal).props().onClose).toBeDefined();

    wrapper.find(Modal).props().onClose();
    expect(setSelectedOrg).toBeCalledWith(null);

    wrapper.find(Modal).find('footer').find(Button).last().simulate('click');
    wrapper.find(Modal).find('textarea').first().simulate('change', 'Super secret');
    wrapper.find(Modal).find('textarea').last().simulate('change', 'Super team');
    expect(wrapper.html()).toMatchSnapshot();
});
