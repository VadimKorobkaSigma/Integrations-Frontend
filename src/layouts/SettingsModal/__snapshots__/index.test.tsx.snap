// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsModal Component 1`] = `"<div class=\\"\\"><div></div><div><header><h1>Settings of test 1 organization</h1></header><form><label for=\\"secret\\">Checkmarx Go Secret</label><textarea placeholder=\\"Secret\\" id=\\"secret\\">Test secret</textarea><label for=\\"team\\">Team</label><textarea placeholder=\\"Team\\" id=\\"team\\">Test team</textarea></form><footer><button variant=\\"link\\" class=\\"\\">Cancel</button><button class=\\"\\">Submit</button></footer></div></div>"`;

exports[`SettingsModal Component 2`] = `"<div class=\\"\\"><div></div><div><header><h1>Settings of test 1 organization</h1></header><form><label for=\\"secret\\">Checkmarx Go Secret</label><textarea placeholder=\\"Secret\\" id=\\"secret\\">Test secret</textarea><label for=\\"team\\">Team</label><textarea placeholder=\\"Team\\" id=\\"team\\">Test team</textarea></form><footer><button variant=\\"link\\" class=\\"\\">Cancel</button><button class=\\"\\">Submit</button></footer></div></div>"`;
